Clippy fixes

ManevilleF opened the pull request:
# Objective

Fixes some `nursery` and `pedantic` clippy warnings like
- `clippy::missing_const_fn`
- `clippy::option_if_let_else`
- `clippy::use_self`
- `clippy::redundant_else`
- `clippy::match_same_arms`
- `clippy::semicolon_if_returned`
- `clippy::explicit_iter_loop`
- `clippy::cloned_instead_of_copied`
- `clippy::map_flatten`

## Solution

I applied many small changes accross all crates.

MinerSebas commented on crates/bevy_ecs/src/storage/blob_vec.rs:
```diff
@@ -44,17 +44,17 @@ impl BlobVec {
-    pub fn len(&self) -> usize {
+    pub const fn len(&self) -> usize {
         self.len
     }
 
     #[inline]
-    pub fn is_empty(&self) -> bool {
+    pub const fn is_empty(&self) -> bool {
         self.len == 0
     }
 
     #[inline]
-    pub fn capacity(&self) -> usize {
+    pub const fn capacity(&self) -> usize {
```
`BlobVec` has no const constructor.

MinerSebas commented on crates/bevy_ecs/src/storage/blob_vec.rs:
```diff
@@ -196,7 +196,7 @@ impl BlobVec {
     /// must ensure rust mutability rules are not violated
     #[inline]
-    pub unsafe fn get_ptr(&self) -> NonNull<u8> {
+    pub const unsafe fn get_ptr(&self) -> NonNull<u8> {
```
Same as above.

MinerSebas commented on crates/bevy_ecs/src/storage/sparse_set.rs:
```diff
@@ -117,12 +117,12 @@ impl ComponentSparseSet {
     #[inline]
-    pub fn len(&self) -> usize {
+    pub const fn len(&self) -> usize {
         self.dense.len()
     }
 
     #[inline]
-    pub fn is_empty(&self) -> bool {
+    pub const fn is_empty(&self) -> bool {
```
`ComponentSparseSet` has no const constructor.

MinerSebas commented on crates/bevy_ecs/src/storage/table.rs:
```diff
@@ -84,12 +84,12 @@ impl Column {
     #[inline]
-    pub fn len(&self) -> usize {
+    pub const fn len(&self) -> usize {
         self.data.len()
     }
 
     #[inline]
-    pub fn is_empty(&self) -> bool {
+    pub const fn is_empty(&self) -> bool {
```
`Column` has no const constructor.

MinerSebas commented on crates/bevy_ecs/src/storage/table.rs:
```diff
@@ -136,7 +136,7 @@ impl Column {
     /// must ensure rust mutability rules are not violated
     #[inline]
-    pub unsafe fn get_data_ptr(&self) -> NonNull<u8> {
+    pub const unsafe fn get_data_ptr(&self) -> NonNull<u8> {
```
Same as above.

MinerSebas commented on crates/bevy_ecs/src/system/commands/mod.rs:
````diff
@@ -372,7 +372,7 @@ impl<'w, 's, 'a> EntityCommands<'w, 's, 'a> {
     /// ```
     #[inline]
-    pub fn id(&self) -> Entity {
+    pub const fn id(&self) -> Entity {
````
`EntityCommands` get created in the `entity()` call of `Commands` which is not const.

MinerSebas commented on crates/bevy_ecs/src/system/function_system.rs:
```diff
@@ -35,7 +35,7 @@ impl SystemMeta {
     /// Returns true if the system is [`Send`].
     #[inline]
-    pub fn is_send(&self) -> bool {
+    pub const fn is_send(&self) -> bool {
         self.is_send
     }
```
`SystemMeta` has no const constructor.

MinerSebas commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -113,12 +113,12 @@ impl<'w> EntityMut<'w> {
     #[inline]
-    pub fn id(&self) -> Entity {
+    pub const fn id(&self) -> Entity {
         self.entity
     }
 
     #[inline]
-    pub fn location(&self) -> EntityLocation {
+    pub const fn location(&self) -> EntityLocation {
```
`EntityMut` has no const constructor.

MinerSebas commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -111,19 +111,19 @@ impl World {
     #[inline]
-    pub fn id(&self) -> WorldId {
+    pub const fn id(&self) -> WorldId {
         self.id
     }
 
     /// Retrieves this world's [Entities] collection
     #[inline]
-    pub fn entities(&self) -> &Entities {
+    pub const fn entities(&self) -> &Entities {
```
`World` has no const constructor.

MinerSebas commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -135,13 +135,13 @@ impl World {
     #[inline]
-    pub fn archetypes(&self) -> &Archetypes {
+    pub const fn archetypes(&self) -> &Archetypes {
         &self.archetypes
     }
 
     /// Retrieves this world's [Components] collection
     #[inline]
-    pub fn components(&self) -> &Components {
+    pub const fn components(&self) -> &Components {
```
Same as above.

MinerSebas commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -153,13 +153,13 @@ impl World {
     #[inline]
-    pub fn storages(&self) -> &Storages {
+    pub const fn storages(&self) -> &Storages {
         &self.storages
     }
 
     /// Retrieves this world's [Bundles] collection
     #[inline]
-    pub fn bundles(&self) -> &Bundles {
+    pub const fn bundles(&self) -> &Bundles {
```
Same as above.

MinerSebas commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -1120,7 +1117,7 @@ impl World {
 
     #[inline]
-    pub fn last_change_tick(&self) -> u32 {
+    pub const fn last_change_tick(&self) -> u32 {
         self.last_change_tick
     }
```
Same as above.

MinerSebas commented on crates/bevy_input/src/touch.rs:
```diff
@@ -98,39 +98,39 @@ impl Touch {
     }
 
     #[inline]
-    pub fn start_position(&self) -> Vec2 {
+    pub const fn start_position(&self) -> Vec2 {
         self.start_position
     }
 
     #[inline]
-    pub fn start_force(&self) -> Option<ForceTouch> {
+    pub const fn start_force(&self) -> Option<ForceTouch> {
         self.start_force
     }
 
     #[inline]
-    pub fn previous_position(&self) -> Vec2 {
+    pub const fn previous_position(&self) -> Vec2 {
         self.previous_position
     }
 
     #[inline]
-    pub fn position(&self) -> Vec2 {
+    pub const fn position(&self) -> Vec2 {
         self.position
     }
 
     #[inline]
-    pub fn force(&self) -> Option<ForceTouch> {
+    pub const fn force(&self) -> Option<ForceTouch> {
```
`Touch` has no const constructor.

MinerSebas commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -174,7 +174,7 @@ impl TypeRegistration {
     /// [`TypeId`]: std::any::TypeId
     #[inline]
-    pub fn type_id(&self) -> TypeId {
+    pub const fn type_id(&self) -> TypeId {
```
`TypeRegistration` has no const constructor.

MinerSebas commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -225,7 +225,7 @@ impl TypeRegistration {
 
     /// Returns the name of the type.
-    pub fn name(&self) -> &'static str {
+    pub const fn name(&self) -> &'static str {
```
Same as above.

MinerSebas commented on crates/bevy_render/src/mesh/mesh/mod.rs:
```diff
@@ -72,15 +72,15 @@ impl Mesh {
 
     /// Returns the topology of the mesh.
-    pub fn primitive_topology(&self) -> PrimitiveTopology {
+    pub const fn primitive_topology(&self) -> PrimitiveTopology {
```
`Mesh` does not have a const constructor.

MinerSebas commented on crates/bevy_render/src/mesh/mesh/mod.rs:
```diff
@@ -116,7 +116,7 @@ impl Mesh {
 
     /// Retrieves the vertex `indices` of the mesh.
-    pub fn indices(&self) -> Option<&Indices> {
+    pub const fn indices(&self) -> Option<&Indices> {
```
Same as above.

MinerSebas commented on crates/bevy_render/src/render_graph/context.rs:
```diff
@@ -48,17 +48,17 @@ impl<'a> RenderGraphContext<'a> {
-    pub fn inputs(&self) -> &[SlotValue] {
+    pub const fn inputs(&self) -> &[SlotValue] {
         self.inputs
     }
 
     /// Returns the [`SlotInfos`] of the inputs.
-    pub fn input_info(&self) -> &SlotInfos {
+    pub const fn input_info(&self) -> &SlotInfos {
         &self.node.input_slots
     }
 
     /// Returns the [`SlotInfos`] of the outputs.
-    pub fn output_info(&self) -> &SlotInfos {
+    pub const fn output_info(&self) -> &SlotInfos {
```
`RenderGraphContext` has no const constructors.

MinerSebas commented on crates/bevy_render/src/render_graph/node.rs:
```diff
@@ -21,10 +21,10 @@ pub struct NodeId(Uuid);
     }
 
-    pub fn uuid(&self) -> &Uuid {
+    pub const fn uuid(&self) -> &Uuid {
```
`NodeId` has no const constructor.

MinerSebas commented on crates/bevy_render/src/render_resource/bind_group.rs:
```diff
@@ -20,14 +20,14 @@ pub struct BindGroup {
     /// Returns the [`BindGroupId`].
     #[inline]
-    pub fn id(&self) -> BindGroupId {
+    pub const fn id(&self) -> BindGroupId {
```
`BindGroup` has no const constructor.

MinerSebas commented on crates/bevy_render/src/render_resource/bind_group_layout.rs:
```diff
@@ -12,7 +12,7 @@ pub struct BindGroupLayout {
 impl BindGroupLayout {
     #[inline]
-    pub fn id(&self) -> BindGroupLayoutId {
+    pub const fn id(&self) -> BindGroupLayoutId {
```
`BindGroupLayout` has no const constructor.

MinerSebas commented on crates/bevy_render/src/render_resource/buffer.rs:
```diff
@@ -15,7 +15,7 @@ pub struct Buffer {
 impl Buffer {
     #[inline]
-    pub fn id(&self) -> BufferId {
+    pub const fn id(&self) -> BufferId {
```
`Buffer` has no const constructor.

MinerSebas commented on crates/bevy_render/src/render_resource/buffer.rs:
```diff
@@ -65,12 +65,12 @@ pub struct BufferSlice<'a> {
     #[inline]
-    pub fn id(&self) -> BufferId {
+    pub const fn id(&self) -> BufferId {
         self.id
     }
 
     #[inline]
-    pub fn offset(&self) -> wgpu::BufferAddress {
+    pub const fn offset(&self) -> wgpu::BufferAddress {
```
`BufferSlice` has no const constructor.

MinerSebas commented on crates/bevy_render/src/render_resource/pipeline.rs:
```diff
@@ -23,14 +23,14 @@ pub struct RenderPipeline {
 impl RenderPipeline {
     #[inline]
-    pub fn id(&self) -> RenderPipelineId {
+    pub const fn id(&self) -> RenderPipelineId {
```
`RenderPipeline` has no const Constructor.

MinerSebas commented on crates/bevy_render/src/render_resource/shader.rs:
```diff
@@ -79,7 +79,7 @@ impl Shader {
 
     #[inline]
-    pub fn import_path(&self) -> Option<&ShaderImport> {
+    pub const fn import_path(&self) -> Option<&ShaderImport> {
```
`Shader` has no const cosntructor.

MinerSebas commented on crates/bevy_render/src/render_resource/texture.rs:
```diff
@@ -18,7 +18,7 @@ pub struct Texture {
     /// Returns the [`TextureId`].
     #[inline]
-    pub fn id(&self) -> TextureId {
+    pub const fn id(&self) -> TextureId {
```
`Texture` has not a const constructor.

MinerSebas commented on crates/bevy_tasks/src/task_pool.rs:
```diff
@@ -31,13 +31,13 @@ impl TaskPoolBuilder {
     /// of logical cores of the system
-    pub fn num_threads(mut self, num_threads: usize) -> Self {
+    pub const fn num_threads(mut self, num_threads: usize) -> Self {
         self.num_threads = Some(num_threads);
         self
     }
 
     /// Override the stack size of the threads created for the pool
-    pub fn stack_size(mut self, stack_size: usize) -> Self {
+    pub const fn stack_size(mut self, stack_size: usize) -> Self {
```
`TaskPoolBuilder` does not have a const-constructor, but the `new` method could be made trivially const.

MinerSebas commented on crates/bevy_window/src/window.rs:
```diff
@@ -436,7 +436,7 @@ impl Window {
 
     #[inline]
-    pub fn resizable(&self) -> bool {
+    pub const fn resizable(&self) -> bool {
```
Same as above.

MinerSebas commented on crates/bevy_window/src/window.rs:
```diff
@@ -447,7 +447,7 @@ impl Window {
 
     #[inline]
-    pub fn decorations(&self) -> bool {
+    pub const fn decorations(&self) -> bool {
```
Same as above.

MinerSebas commented on crates/bevy_window/src/window.rs:
```diff
@@ -458,7 +458,7 @@ impl Window {
 
     #[inline]
-    pub fn cursor_locked(&self) -> bool {
+    pub const fn cursor_locked(&self) -> bool {
```
Same as above.

MinerSebas commented on crates/bevy_window/src/window.rs:
```diff
@@ -469,7 +469,7 @@ impl Window {
 
     #[inline]
-    pub fn cursor_visible(&self) -> bool {
+    pub const fn cursor_visible(&self) -> bool {
```
Same as above.

MinerSebas commented on crates/bevy_window/src/window.rs:
```diff
@@ -481,7 +481,7 @@ impl Window {
 
     #[inline]
-    pub fn cursor_icon(&self) -> CursorIcon {
+    pub const fn cursor_icon(&self) -> CursorIcon {
```
Same as above.

MinerSebas commented on crates/bevy_window/src/window.rs:
```diff
@@ -492,7 +492,7 @@ impl Window {
     /// The current mouse position, in physical pixels.
     #[inline]
-    pub fn physical_cursor_position(&self) -> Option<DVec2> {
+    pub const fn physical_cursor_position(&self) -> Option<DVec2> {
```
Same as above.

MinerSebas commented on crates/bevy_window/src/window.rs:
```diff
@@ -522,7 +522,7 @@ impl Window {
 
     #[inline]
-    pub fn mode(&self) -> WindowMode {
+    pub const fn mode(&self) -> WindowMode {
```
Same as above.

MinerSebas commented on crates/bevy_window/src/window.rs:
```diff
@@ -540,7 +540,7 @@ impl Window {
 
     #[inline]
-    pub fn is_focused(&self) -> bool {
+    pub const fn is_focused(&self) -> bool {
```
Same as above.

MinerSebas commented on crates/bevy_core/src/time/stopwatch.rs:
```diff
@@ -60,7 +60,7 @@ impl Stopwatch {
     /// [`elapsed_secs`](Stopwatch::elapsed) - if a `f32` value is desirable instead.
     #[inline]
-    pub fn elapsed(&self) -> Duration {
+    pub const fn elapsed(&self) -> Duration {
```
`Stopwatch` has no const constructor.

MinerSebas commented on crates/bevy_core/src/time/stopwatch.rs:
````diff
@@ -168,7 +168,7 @@ impl Stopwatch {
     /// ```
     #[inline]
-    pub fn paused(&self) -> bool {
+    pub const fn paused(&self) -> bool {
````
Same as above.

MinerSebas commented on crates/bevy_core/src/time/time.rs:
```diff
@@ -47,43 +47,43 @@ impl Time {
 
     /// The delta between the current and last tick as [`f32`] seconds
     #[inline]
-    pub fn delta_seconds(&self) -> f32 {
+    pub const fn delta_seconds(&self) -> f32 {
         self.delta_seconds
     }
 
     /// The delta between the current and last tick as [`f64`] seconds
     #[inline]
-    pub fn delta_seconds_f64(&self) -> f64 {
+    pub const fn delta_seconds_f64(&self) -> f64 {
         self.delta_seconds_f64
     }
 
     /// The time from startup to the last update in seconds
     #[inline]
-    pub fn seconds_since_startup(&self) -> f64 {
+    pub const fn seconds_since_startup(&self) -> f64 {
         self.seconds_since_startup
     }
 
     /// The [`Instant`] the app was started
     #[inline]
-    pub fn startup(&self) -> Instant {
+    pub const fn startup(&self) -> Instant {
         self.startup
     }
 
     /// The ['Instant'] when [`Time::update`] was last called, if it exists
     #[inline]
-    pub fn last_update(&self) -> Option<Instant> {
+    pub const fn last_update(&self) -> Option<Instant> {
         self.last_update
     }
 
     /// The ['Duration'] from startup to the last update
     #[inline]
-    pub fn time_since_startup(&self) -> Duration {
+    pub const fn time_since_startup(&self) -> Duration {
```
`Time` has no const constructor.

MinerSebas commented on crates/bevy_core/src/time/timer.rs:
````diff
@@ -60,7 +60,7 @@ impl Timer {
     /// ```
     #[inline]
-    pub fn finished(&self) -> bool {
+    pub const fn finished(&self) -> bool {
````
`Timer` has no non-const constructor.

MinerSebas commented on crates/bevy_core/src/time/timer.rs:
````diff
@@ -77,7 +77,7 @@ impl Timer {
     /// ```
     #[inline]
-    pub fn just_finished(&self) -> bool {
+    pub const fn just_finished(&self) -> bool {
````
Same as above.

MinerSebas commented on crates/bevy_core/src/time/timer.rs:
````diff
@@ -95,7 +95,7 @@ impl Timer {
     /// ```
     #[inline]
-    pub fn elapsed(&self) -> Duration {
+    pub const fn elapsed(&self) -> Duration {
````
Same as above.

MinerSebas commented on crates/bevy_core/src/time/timer.rs:
````diff
@@ -135,7 +135,7 @@ impl Timer {
     /// ```
     #[inline]
-    pub fn duration(&self) -> Duration {
+    pub const fn duration(&self) -> Duration {
````
Same as above.

MinerSebas commented on crates/bevy_core/src/time/timer.rs:
````diff
@@ -163,7 +163,7 @@ impl Timer {
     /// ```
     #[inline]
-    pub fn repeating(&self) -> bool {
+    pub const fn repeating(&self) -> bool {
````
Same as above.

MinerSebas commented on crates/bevy_core/src/time/timer.rs:
````diff
@@ -285,7 +285,7 @@ impl Timer {
     /// ```
     #[inline]
-    pub fn paused(&self) -> bool {
+    pub const fn paused(&self) -> bool {
````
Same as above.

MinerSebas commented on crates/bevy_core/src/time/timer.rs:
````diff
@@ -359,7 +359,7 @@ impl Timer {
     /// ```
     #[inline]
-    pub fn times_finished(&self) -> u32 {
+    pub const fn times_finished(&self) -> u32 {
````
Same as above.

MinerSebas reviewed (changes requested):
There were several places where an `fn` was made const, that can't be actually called in a const context as the Type they are implemented for can only be constructed at runtime.
Please revert those functions or try to get a const-constructor.

Also please when change as many Files/Places as here, make 1. Commit for every clippy Lint. this makes it much easier to review each individual Lint.

ManevilleF commented:
@MinerSebas `const` functions don't require a `const` context, there is no need for `const` constructors as it just allows the evaluation of the function at compile time. 
Is there maybe a convention regarding `const` functions that I'm not aware of?

mockersf commented:
A `const` function can be evaluated at compile time, and will if all its parameters are `const`. In the cases raised by MinerSebas, there will never be a `self` that is `const`, so even if the functions are marked, they will never be evaluated at compile time.

Adding `const` here won't change anything for now, but it's an extra constraint on how Bevy may evolve as they may need to not be `const` later on and the person may be hesitant to remove it.

alice-i-cecile commented:
Some of this work was duplicated in #2934. I think I prefer the clarity and freshness of this PR, if we're picking one to merge.

ManevilleF commented:
Thanks @mockersf (Merci François !)  I'll apply the suggestions of @MinerSebas and remove the extra `const` I added

@alice-i-cecile I added two more clippy fixes: `clippy::semicolon_if_nothing_is_returned` and `clippy::explicit_iter_loop` which are in the similar PR.

Should I add to the CI config or the `lib.rs` files the now fixed clippy warnings to avoid them in the future?

ManevilleF commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -570,6 +570,7 @@ impl World {
 
     /// Returns an iterator of entities that had components of type `T` removed
     /// since the last call to [`World::clear_trackers`].
+    // TODO: return `std::iter::Copied` instead
```
I think we should return a `std::iter::Copied` here but I don't know if it would be a breaking change or not

ManevilleF commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -579,6 +580,7 @@ impl World {
 
     /// Returns an iterator of entities that had components with the given `component_id` removed
     /// since the last call to [`World::clear_trackers`].
+    // TODO: return `std::iter::Copied` instead
```
I think we should return a `std::iter::Copied` here but I don't know if it would be a breaking change or not

alice-i-cecile commented:
Yeah, if we're going to fix pedantic / nursery lints we should toggle them on in CI so they stay fixed.

mockersf commented:
> Yeah, if we're going to fix pedantic / nursery lints we should toggle them on in CI so they stay fixed.

To be transparent, this is exactly why I think most of those lints should not be fixed. adding more things to control while slow down development. Some add enough values to be worth it, but that's not the case of most of the pedantic lints.

ManevilleF commented:
@mockersf I agree, many pedantic lints are not worth it.
I only fixed a few that I think are good practice or useful and ignored the others.

alice-i-cecile commented:
@ManevilleF do you want to rebase this? If so we can merge it in ASAP after rebasing; otherwise I think we should just close it out to reduce merge conflicts.

ManevilleF commented:
As discussed I will close this and open it again from the `main` branch with the warnings enabled on the CI.

@mockersf @cart I think the current list of warnings is a good start:

```
    clippy::option_if_let_else
    clippy::use_self
    clippy::redundant_else
    clippy::match_same_arms
    clippy::semicolon_if_returned
    clippy::explicit_iter_loop
    clippy::cloned_instead_of_copied
    clippy::map_flatten
```

Do you have more warnings you wish to add or some to remove?

alice-i-cecile commented:
[List of lints](https://rust-lang.github.io/rust-clippy/v0.0.212/index.html), for easy reference.

alice-i-cecile commented:
Personal opinions:

1. [option_if_let_else](https://rust-lang.github.io/rust-clippy/master/#option_if_let_else): genuinely clearer.
2. [use_self](https://rust-lang.github.io/rust-clippy/master/#use_self): actively dislike this. See #4152.
3. [redundant_else](https://rust-lang.github.io/rust-clippy/master/#redundant_else): seems nice
4. [match_same_arms](https://rust-lang.github.io/rust-clippy/master/#match_same_arms): nice clarity win
5. [semicolon_if_nothing_returned](https://rust-lang.github.io/rust-clippy/master/#semicolon_if_nothing_returned): presuming this is what you meant? Seems like a minor improvement.
6. [explicit_iter_loop](https://rust-lang.github.io/rust-clippy/master/#explicit_iter_loop): no strong feelings
7. [cloned_instead_of_copied](https://rust-lang.github.io/rust-clippy/master/#cloned_instead_of_copied): eh sure why not
8. [map_flatten](https://rust-lang.github.io/rust-clippy/master/#map_flatten): I like it! Much clearer.
